### DISCUSSION TITLE
Different cache databases available

### DISCUSSION THREAD

**Discussion by afbjorklund** [ccache/ccache#900] [Development]
We had some discussions in the various storage backends, about the different databases:

~https://github.com/ccache/ccache/pull/58~

~https://github.com/ccache/ccache/issues/859~

So far the plan is to _replace_ memcached with redis, both for popularity and for simplicity...

Think I will leave Couchbase and Redis Enterprise out of it, and only look at Open Source.

Investigating RocksDB as a backend (for Redis), as well as looking at Kyoto Tycoon/Cabinet.

https://github.com/ccache/ccache/issues/422

https://github.com/ccache/ccache/issues/435

![db-engines-ranking-ccache](https://user-images.githubusercontent.com/10364051/125835814-7a1c60d6-c127-4fa2-bdc3-de73a3a92847.png)

Ranking from https://db-engines.com/en/ranking

----

I would be interested in feedback on:

1) **Is it OK to "only" do a Redis backend, and **not** do a Memcached backend ?**

So far: **Yes**, it can replace it

2) **Is there some _other_ database that should be included by default (e.g. KT)**

So far: **No**, not at this moment

**Comment by afbjorklund** [comment]
I suppose one _could_ also discuss different network file systems and different web servers, but I wanted some input on databases...

~FileStorage~

~HttpStorage~

Would it be enough with Redis protocol, or should there be more backends ? Should ccache even have a special cache server ?

RedisStorage

KyotoStorage (?)

**Comment by afbjorklund** [comment]
More background information, as in 2010-2020: (note that ccache is 20 years old itself)

https://stackoverflow.com/questions/23601622/if-redis-is-already-a-part-of-the-stack-why-is-memcached-still-used-alongside-r

https://stackoverflow.com/questions/2873249/is-memcached-a-dinosaur-in-comparison-to-redis

As noted elsewhere, it is possible to do workarounds and alternatives for Redis being single-threaded and single-server.

----

It is possible to use built-in "hybrids" (i.e. between ram and ssd), such as Memcached's "extstore" and Redis-on-Flash etc.

But mostly we looked at alternative servers such as [memcachedb](https://github.com/afbjorklund/memcachedb) and [ardb](https://github.com/afbjorklund/ardb), for that particular scenario (memory-mapped)

Another important aspect for distributed performance is running a local proxy before the network, such as moxi or nutcracker.

Most casual users of ccache will **not** bother with this, but just use the built-in file and http support. Which could need some docs.

**Comment by afbjorklund** [comment]
There _was_ a backend for MySQL (InnoDB), but it was not included here. They suggest to use memcached, if you want performance.

The reason for writing it was that some environments only know how to host SQL databases, and have no idea about NoSQL...

**Comment by jrosdahl** [comment]
Hi Anders,

Thanks for the discussion topic.

I don't know what to answer, really, since scaling networked databases is not my area of expertise, and I don't have bandwidth to investigate the options. However, one thing I do have experience with is software development and maintenance, and I'm not very interested in having a multitude of backends in ccache for maintainability reasons. This is a consequence of the decision not to implement a [plugin framework](https://github.com/ccache/ccache/issues/414).

I also know that I never heard of Kyoto Tycoon before #435, so I would like to have some convincing indications that it's worth bothering about having a special backend for the binary protocol. (If I understand correctly it also provides an HTTP REST API compatible with our HTTP storage backend. Haven't tried, though.)

> Would it be enough with Redis protocol, or should there be more backends ?

Don't know, but I think that it's reasonable to start with HTTP+Redis. I suspect that various cloud storages could also be of interest.

> Should ccache even have a special cache server ?

Don't know, but it spontaneously does not seem like a very interesting idea unless it solves some problem that is hard to solve another way.

----

Just some notes of clarification in case somebody else reads the discussion:

> So far the plan is to replace memcached with redis, both for popularity and for simplicity...

"replace memcached" here means "not add memcached support".

> But mostly we looked at [...]
> There _was_ a backend for MySQL (InnoDB) [...]

I think that the above statements refer to work done by @afbjorklund and colleagues at some previous company. It doesn't mean "we" as in "the ccache project".

@afbjorklund: Correct me if I misunderstood something. :slightly_smiling_face:

**Reply by afbjorklund** [reply]
The MySQL backend PoC was here: b6504277ef66d7f5825d7204cafb3b64e045d570 - I just hacked some different backends together, another one was Redis...

```SQL
CREATE TABLE IF NOT EXISTS `ccache` (
                           `key` varchar(42) NOT NULL, 
                           `val` longblob, 
                           `ts` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP, 
                           PRIMARY KEY (`key`));

REPLACE INTO `ccache` (`key`, `val`) VALUES (?, ?);

SELECT `val` FROM `ccache` WHERE `key` = ?;
```

It was lost a bit in the discussion on the HTTP backend. But my experience of "IT database hosting" was from work, yes :-)

**Reply by afbjorklund** [reply]
> I also know that I never heard of Kyoto Tycoon before #435

The main feature of Kyoto Tycoon (_slightly_ more wellknown from the predecessor Tokyo Tyrant) was the persistence feature.
But since Redis has "disk backup" (disaster recovery) built-in, and "memory mapped" (hybrid mode) available it is not unique.

There was a nice summary here: https://charlesleifer.com/blog/kyoto-tycoon-in-2019/

The code is readable and nice, and it "works". But I'm not sure that it is notable enough ?

When going with a standard option like Redis, there are packages available in the distribution and support + documentation.
This one needs to be built from source, and even patched/down-stream at that. And I'm not seeing any killer features...

The main reason for doing a Proof-of-Concept for it, was so that we can run some benchmarks.

I think it will come out slower than the other options, but it would be nice to have some numbers...

**Reply by afbjorklund** [reply]
> f I understand correctly it also provides an HTTP REST API compatible with our HTTP storage backend. 

I think it uses a custom protocol, basically calls under `/rpc` while using TSV instead of something else (XML?).

https://dbmx.net/kyototycoon/spex.html#protocol

But the HTTP overhead as such should be similar, when compared to some lower-level TCP custom protocol.

https://redis.io/topics/protocol

**Reply by afbjorklund** [reply]
Most of my experience is from on-prem datacenters, with plenty of servers and each with plenty of memory available.
Therefore it was natural to go with the fastest and most scalable option (memcached), even if harder to implement etc.

The movement towards more disk-based (memory-mapped) approaches is for the cloud, which has a different cost model.
I think using HTTP will be the best "out-of-the-box", and then use a Redis-based approach when wanting to improve ?

**Reply by jrosdahl** [reply]
> I think it uses a custom protocol, basically calls under `/rpc` while using TSV instead of something else (XML?).

See <https://dbmx.net/kyototycoon/spex.html#protocol_restful>.

**Reply by afbjorklund** [reply]
I actually just used the provided library, and didn't experiment with the REST (or the memcached "pluggable server")

**Reply by afbjorklund** [reply]
Closed the KT PR #893, seems like it won't need the custom storage backend - if anyone will actually be using it at all.

**Comment by afbjorklund** [comment]
Currently I'm thinking to leave these out of it: Memcached, MySQL, Kyoto Tycoon...

And just go with these as the default backends: File, HTTP (+SSL?), Redis (+SSL?)

That "should be enough", and as they have a reasonable implementation already...

Long-term, I would like to see "something else" such as https://github.com/infinit/memo

----

<s>
Or perhaps some more traditional building block like RADOS: [https://docs.ceph.com/](https://docs.ceph.com/en/latest/architecture/)

`storage::secondary::RadosStorage` <https://docs.ceph.com/en/latest/rados/api/librados/>

But I think such systems can be experimented with already using File backend and FUSE ?

* https://docs.ceph.com/en/latest/cephfs/index.html (available both as kernel and fuse, actually)
> CephFS, is a POSIX-compliant file system built on top of Ceph’s distributed object store, RADOS. 
</s>

----

<s>
The RocksDB solution (ARDB) looks solid too, and doesn't require _any_ changes to the Redis backend.

That is, the hiredis client library can be used with all the different servers implementing the Redis protocol:

* https://github.com/yinqiwen/ardb
* https://github.com/ideawu/ssdb
* https://github.com/EQ-Alpha/KeyDB
* https://github.com/Tencent/Tendis
</s>

**Reply by afbjorklund** [reply]
I opened #904 if someone wants to investigate using Ceph, like for a future addition (i.e. not 4.4)

We need to have some benchmarks in place, that would motivate the storage decisions further ?

**Reply by afbjorklund** [reply]
As noted below, details on Redis/RocksDB are moved to the wiki instead of being updated here.

Having _one_ option to extend storage to disk would be enough, but currently not clear which one...

**Comment by jrosdahl** [comment]
I've now created two wiki pages for the existing storage backends:

* https://github.com/ccache/ccache/wiki/HTTP-storage
* https://github.com/ccache/ccache/wiki/Redis-storage

They will in turn (indirectly) be referenced from the ccache manual. I'll maybe convert the wiki pages to web pages on https://github.com/ccache/ccache.github.io instead in the future, but for now we can collect ideas and recommendations in a more light-weight way.

@afbjorklund: If you want to, it would be great if you could help out and share your knowledge and experience, especially for the Redis part. I added the alternatives you mentioned above (Ardb, SSDB, KeyDB, Tendis) to the wiki page, but I don't think we need to document all of them in detail (or even keep references to them) if we don't want to recommend using them.

**Reply by afbjorklund** [reply]
Sounds good, that is probably plenty of backends to start with... I think I will close this discussion.

There doesn't seem to be one _obvious_ candidate between the Redis alternatives, so one would have to try them and see (like: pros and cons). The default option for ccache would be to run the "regular" Redis server, which is limited to available memory. I do think RocksDB is looking good (same as [Redis-on-Flash](https://docs.redislabs.com/latest/rs/concepts/memory-architecture/redis-flash/) is using*), but there are some issues with maintenance and packaging...

\* according to https://redislabs.com/blog/hood-redis-enterprise-flash-database-architecture/ at least

The wiki will probably just describe the general principle about the memory-mapped option, and then offer links to the various Open Source projects. Perhaps offer a generic link to Redis Labs and Redis Enterprise as well, for completeness. But I would expect all of them to have similar I/O performance characteristics, when compared to (in-memory) Redis and some (on-disk) HTTP.

**Comment by afbjorklund** [comment]
> Investigating RocksDB as a backend (for Redis), as well as looking at Kyoto Tycoon/Cabinet.

Tried gRPC (instead of the Redis protocol), and the successor to KT called Tkrzw ("TK" for short).

In theory, it (gRPC over HTTP/2) should be somewhere in between the REST and Redis protocols ?

And the disk persistance and replication was provided by the server, but there was no proxy provided.

**Comment by afbjorklund** [comment]
Here is another one, called Dragonfly DB: https://dragonflydb.io/

Discussion at: https://news.ycombinator.com/item?id=31560547
